base/compiler/abstractinterpretation.jl

```diff
@@ -1312,19 +1340,21 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
                 l = stmt.dest::Int
                 # constant conditions
                 if condval === true
+                    if isa(condt, Conditional)
+                        changes = conditional_changes(changes, condt.vtype, condt.var)
+                    end
                 elseif condval === false
                     pc´ = l
+                    if isa(condt, Conditional)
+                        changes = conditional_changes(changes, condt.elsetype, condt.var)
+                    end
```

Comment:
Constant conditions can be `Conditional` and convey an additional constraint, and these changes will propagate it as well as the "general case" `Conditional`s.

E.g. the following example will be improved:
```julia
f(a) = g(a) ? a : nothing
g(a::Int) = true

Base.return_types(f, (Any,)) # ::Int
```

Reply:
Yes, these really should have an even stronger effect than that and act like a type-assert on the statement, which then constrains `a` on every subsequent statement. We can't handle that general case yet, but for now, it might be more accurate to put this logic into `stupdate!` to occur after every statement, instead of being specific to branches. I'm inclined to say we should omit this particular change, because it is largely unrelated to this PR, and is unable to handle many other cases.

```
f(a) = (g(a); a)
g(a::Int) = true # ::Conditional(SlotNumber(2), Int, Union{}), constant conditional constraint

@test Base.return_types(f, (Any,)) == Any[Int]
```

Reply:
thanks, I agree with you, let's improve this in higher level (reverted in 6b31e2d)